// tdesign-vue-next/cjs/anchor/anchor-item.js
/**
 * tdesign v1.19.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var anchor_utils_index = require('./utils/index.js');
var anchor_anchorItemProps = require('./anchor-item-props.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-56b80cd8.js');
var index$1 = require('../_chunks/dep-33c128cc.js');
var index = require('../_chunks/dep-0864235d.js');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-119f7442.js');
var anchor_constants_index = require('./constants/index.js');
var isString = require('../_chunks/dep-df3fb11d.js');
var isFunction = require('../_chunks/dep-d866b4dd.js');
require('../_chunks/dep-29d9a122.js');
require('../config-provider/hooks/useConfig.js');
require('../_chunks/dep-f6b59916.js');
require('../_chunks/dep-ade706c5.js');
require('dayjs');
require('../_chunks/dep-1d717c89.js');
require('../_chunks/dep-fda07f7d.js');
require('../_chunks/dep-dca49e22.js');
require('../_chunks/dep-0f042eb2.js');
require('../_chunks/dep-8ff8ef8e.js');
require('../_chunks/dep-477a86d3.js');
require('../_chunks/dep-0ff4bc31.js');
require('../_chunks/dep-420f2c79.js');
require('../_chunks/dep-5e8266b1.js');
require('../_chunks/dep-72c5ba77.js');
require('../_chunks/dep-fdd955c0.js');
require('../_chunks/dep-63b15b94.js');
require('../_chunks/dep-b40cc19a.js');
require('../_chunks/dep-be7c003a.js');
require('../_chunks/dep-5c7e4acc.js');
require('../_chunks/dep-00e50b7c.js');
require('../_chunks/dep-332d4a09.js');
require('../_chunks/dep-f6087cc4.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var localProps = _objectSpread(_objectSpread({}, anchor_anchorItemProps["default"]), {}, {
  href: {
    type: String,
    required: true,
    validator: function validator(v) {
      return anchor_utils_index.ANCHOR_SHARP_REGEXP.test(v);
    }
  }
});
var _AnchorItem = Vue.defineComponent({
  name: "TAnchorItem",
  inject: {
    tAnchor: {
      "default": void 0
    }
  },
  props: localProps,
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var anchor = Vue.inject(anchor_constants_index.AnchorInjectionKey, void 0);
    var CLASSNAME_PREFIX = index.usePrefixClass("anchor__item");
    var _useCommonClassName = index$1.useCommonClassName(),
      STATUS = _useCommonClassName.STATUS;
    var register = function register() {
      anchor.registerLink(props2.href);
    };
    var unregister = function unregister(hrefVal) {
      var val = hrefVal || props2.href;
      if (!val) return;
      anchor.unregisterLink(val);
    };
    var handleClick = function handleClick(e) {
      var href = props2.href,
        title = props2.title;
      if (!props2.customScroll) anchor.handleScrollTo(href);
      anchor.handleLinkClick({
        href: href,
        title: isString.isString(title) ? title : void 0,
        e: e
      });
    };
    var renderTitle = function renderTitle() {
      var title = props2.title;
      var titleSlot = slots.title;
      var titleVal;
      if (isString.isString(title)) {
        titleVal = title;
      } else if (isFunction.isFunction(title)) {
        titleVal = title(Vue.h);
      } else if (titleSlot) {
        titleVal = titleSlot(null);
      }
      return titleVal;
    };
    Vue.watch(function () {
      return props2.href;
    }, function (newVal, oldVal) {
      unregister(oldVal);
      register();
    }, {
      immediate: true
    });
    Vue.onMounted(function () {
      register();
    });
    Vue.onUnmounted(function () {
      unregister();
    });
    return function () {
      var href = props2.href,
        target = props2.target;
      var children = slots["default"],
        titleSlot = slots.title;
      var title = renderTitle();
      var titleAttr = isString.isString(title) ? title : null;
      var active = anchor.active === href;
      var wrapperClass = _defineProperty__default["default"](_defineProperty__default["default"]({}, CLASSNAME_PREFIX.value, true), STATUS.value.active, active);
      var titleClass = _defineProperty__default["default"]({}, "".concat(CLASSNAME_PREFIX.value, "-link"), true);
      return Vue.createVNode("div", {
        "class": wrapperClass
      }, [Vue.createVNode("a", {
        "href": href,
        "title": titleAttr,
        "class": titleClass,
        "target": target,
        "onClick": handleClick
      }, [titleSlot ? titleSlot(null) : title]), children && children(null)]);
    };
  }
});

exports["default"] = _AnchorItem;
